cmds/core/dirname/dirname.go

```diff
@@ -9,17 +9,34 @@ package main
 
 import (
 	"fmt"
+	"io"
 	"log"
 	"os"
 	"path/filepath"
 )
 
-func main() {
-	if len(os.Args) < 2 {
-		log.Fatalf("dirname: missing operand")
+func runDirname(w io.Writer, dirname string, args []string) error {
+	if dirname == "" {
+		return fmt.Errorf("dirname: missing operand")
+	}
+	fmt.Fprintf(w, "%s\n", filepath.Dir(dirname))
+	for _, n := range args {
+		if _, err := fmt.Fprintf(w, "%s\n", filepath.Dir(n)); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	for _, n := range os.Args[1:] {
-		fmt.Println(filepath.Dir(n))
+func main() {
+	if len(os.Args) == 0 {
+		log.Fatal()
+	}
+	dirname := ""
+	if len(os.Args[1:]) > 1 {
+		dirname = os.Args[1]
+	}
+	if err := runDirname(os.Stdout, dirname, os.Args[2:]); err != nil {
```

Comment:
so the rules of dirname make everything harder. 
Could we do this:
switch(len(os.Args) {
case 0: error
case 1: dirname, args = os.Args[1], os.Args[1:]
default: dirname, args = os.Args[1], os.Args[2:]
}

or some such.